pkg/yurthub/proxy/remote/loadbalancer.go

```diff
@@ -163,10 +184,170 @@ func (lb *loadBalancer) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+}
+
+func (lb *loadBalancer) modifyResponse(resp *http.Response) error {
+	if resp == nil || resp.Request == nil {
+		klog.Infof("no request info in response, skip cache response")
+		return nil
+	}
+
+	req := resp.Request
+	ctx := req.Context()
+
+	// re-added transfer-encoding=chunked response header for watch request
+	info, exists := apirequest.RequestInfoFrom(ctx)
+	if exists {
+		if info.Verb == "watch" {
+			klog.V(5).Infof("add transfer-encoding=chunked header into response for req %s", hubutil.ReqString(req))
+			h := resp.Header
+			if hv := h.Get("Transfer-Encoding"); hv == "" {
+				h.Add("Transfer-Encoding", "chunked")
+			}
+		}
+	}
+
+	if resp.StatusCode >= http.StatusOK && resp.StatusCode <= http.StatusPartialContent {
+		// prepare response content type
+		reqContentType, _ := hubutil.ReqContentTypeFrom(ctx)
+		respContentType := resp.Header.Get("Content-Type")
+		if len(respContentType) == 0 {
+			respContentType = reqContentType
+		}
+		ctx = hubutil.WithRespContentType(ctx, respContentType)
+		req = req.WithContext(ctx)
+
+		// filter response data
+		if lb.filterManager != nil {
+			if ok, runner := lb.filterManager.FindRunner(req); ok {
+				wrapBody, needUncompressed := hubutil.NewGZipReaderCloser(resp.Header, resp.Body, req, "filter")
+				size, filterRc, err := runner.Filter(req, wrapBody, lb.stopCh)
+				if err != nil {
+					klog.Errorf("failed to filter response for %s, %v", hubutil.ReqString(req), err)
+					return err
+				}
+				resp.Body = filterRc
+				if size > 0 {
+					resp.ContentLength = int64(size)
+					resp.Header.Set("Content-Length", fmt.Sprint(size))
+				}
+
+				// after gunzip in filter, the header content encoding should be removed.
+				// because there's no need to gunzip response.body again.
+				if needUncompressed {
+					resp.Header.Del("Content-Encoding")
+				}
+			}
+		}
+
+		if lb.workingMode == hubutil.WorkingModeEdge {
+			// cache resp with storage interface
+			lb.cacheResponse(req)
+		}
+	} else if resp.StatusCode == http.StatusNotFound && info.Verb == "list" && lb.localCacheMgr != nil {
+		// 404 Not Found: The CRD may have been unregistered and should be updated locally as well.
+		// Other types of requests may return a 404 response for other reasons (for example, getting a pod that doesn't exist).
+		// And the main purpose is to return 404 when list an unregistered resource locally, so here only consider the list request.
+		gvr := schema.GroupVersionResource{
+			Group:    info.APIGroup,
+			Version:  info.APIVersion,
+			Resource: info.Resource,
+		}
+
+		err := lb.localCacheMgr.DeleteKindFor(gvr)
+		if err != nil {
+			klog.Errorf("failed: %v", err)
+		}
+	}
+	return nil
+}
+
+func (lb *loadBalancer) cacheResponse(req *http.Request) {
+	if lb.localCacheMgr.CanCacheFor(req) {
+		ctx := req.Context()
+		poolCacheManager, isHealthy := lb.coordinator.IsHealthy()
+
+		if isHealthy && poolCacheManager != nil {
+			if !isLeaderHubUserAgent(ctx) && isPoolScopedCtx(ctx) {
```

Comment:
I think node and pods for kubelet also should be cached into pool-coordinator.

Reply:
fixed